File: HealthApp/Tabs/Home/Main/Coordinator/HomeTabCoordinator.m
//
//  HomeTabCoordinator.m
//  HealthApp
//
//  Created by Christopher Ray on 3/27/21.
//

#import "HomeTabCoordinator.h"
#import "AppDelegate.h"
#import "AddWorkoutCoordinator.h"
#import "HomeViewController.h"
#import "HomeSetupWorkoutModalViewController.h"
#include "AppUserData.h"
#include "CalendarDateHelpers.h"

#define UpdateGreeting 0x1
#define ResetWorkouts 0x2
#define UpdateWorkouts 0x4

typedef enum {
    CustomWorkoutIndexTestMax,
    CustomWorkoutIndexEndurance,
    CustomWorkoutIndexStrength,
    CustomWorkoutIndexSE,
    CustomWorkoutIndexHIC
} CustomWorkoutIndex;

static inline HomeViewController *getHomeViewController(UINavigationController *controller) {
    return (HomeViewController *) controller.viewControllers[0];
}

void updateUI(HomeTabCoordinator *this, unsigned char updates) {
    HomeViewController *homeVC = getHomeViewController(this->navigationController);
    if (updates & ResetWorkouts) {
        homeViewModel_fetchData(&this->viewModel);
        [homeVC createWorkoutsList];
    }
    if (updates & UpdateWorkouts) {
        [homeVC updateWorkoutsList];
    }
}

void navigateToAddWorkout(HomeTabCoordinator *this, bool dismissVC, Workout *workout) {
    if (dismissVC) [this->navigationController.viewControllers[0]
                    dismissViewControllerAnimated:true completion:nil];
    AddWorkoutCoordinator *child = malloc(sizeof(AddWorkoutCoordinator));
    if (!child) return;

    child->navigationController = this->navigationController;
    child->parent = this;
    child->viewModel.workout = workout;
    this->childCoordinator = child;
    addWorkoutCoordinator_start(child);
}

void homeCoordinator_free(HomeTabCoordinator *this) {
    if (this->childCoordinator) addWorkoutCoordinator_free(this->childCoordinator);
    homeViewModel_free(&this->viewModel);
    free(this);
}

void homeCoordinator_start(HomeTabCoordinator *this) {
    homeViewModel_init(&this->viewModel);
    HomeViewController *vc = [[HomeViewController alloc] initWithDelegate:this];
    [this->navigationController setViewControllers:@[vc]];
    [vc release];
}

void homeCoordinator_didFinishAddingWorkout(HomeTabCoordinator *this, int totalCompletedWorkouts) {
    HomeViewController *homeVC = getHomeViewController(this->navigationController);
    [homeVC updateWorkoutsList];

    const bool showConfetti = homeViewModel_shouldShowConfetti(&this->viewModel,
                                                               totalCompletedWorkouts);

    addWorkoutCoordinator_free(this->childCoordinator);
    this->childCoordinator = NULL;
    [this->navigationController popViewControllerAnimated:true];

    if (showConfetti) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, NSEC_PER_SEC * 0.75),
                       dispatch_get_main_queue(), ^ (void) {
            [homeVC showConfetti];
        });
    }
}

void homeCoordinator_addWorkoutFromPlan(HomeTabCoordinator *this, int index) {
    unsigned char plan = (unsigned char) appUserDataShared->currentPlan;
    Workout *w = exerciseManager_getWeeklyWorkoutAtIndex(plan, appUserData_getWeekInPlan(), index);
    if (w) navigateToAddWorkout(this, false, w);
}

void homeCoordinator_addWorkoutFromCustomButton(HomeTabCoordinator *this, int index) {
    unsigned char type = WorkoutTypeStrength;
    switch (index) {
        case CustomWorkoutIndexSE:
            type = WorkoutTypeSE;
            break;
        case CustomWorkoutIndexHIC:
            type = WorkoutTypeHIC;
            break;
        case CustomWorkoutIndexTestMax: ;
            Workout *w = exerciseManager_getWorkoutFromLibrary(WorkoutTypeStrength, 2, 1, 1, 100);
            if (w) navigateToAddWorkout(this, false, w);
            return;
        case CustomWorkoutIndexEndurance:
            type = WorkoutTypeEndurance;
        default:
            break;
    }

    int count = 0;
    CFStringRef *names = exerciseManager_getWorkoutNamesForType(type, &count);
    if (!names) return;
    else if (!count) {
        free(names);
        return;
    }

    UIViewController *modal = [[HomeSetupWorkoutModalViewController alloc]
                               initWithDelegate:this type:type names:names count:count];
    UINavigationController *container = [[UINavigationController alloc]
                                         initWithRootViewController:modal];
    [this->navigationController.viewControllers[0] presentViewController:container animated:true
                                                              completion:nil];
    [container release];
    [modal release];
}

void homeCoordinator_finishedSettingUpCustomWorkout(HomeTabCoordinator *this, unsigned char type,
                                                    int index, int sets, int reps, int weight) {
    Workout *w = exerciseManager_getWorkoutFromLibrary(type, index, reps, sets, weight);
    if (w) navigateToAddWorkout(this, true, w);
}

void homeCoordinator_checkForChildCoordinator(HomeTabCoordinator *this) {
    AddWorkoutCoordinator *child = this->childCoordinator;
    if (child) {
        addWorkoutCoordinator_stopWorkoutFromBackButtonPress(child);
        this->childCoordinator = NULL;
    }
}

void homeCoordinator_resetUI(HomeTabCoordinator *this) {
    HomeViewController *homeVC = getHomeViewController(this->navigationController);
    homeViewModel_fetchData(&this->viewModel);
    [homeVC createWorkoutsList];
}

void homeCoordinator_updateUI(HomeTabCoordinator *this) {
    HomeViewController *homeVC = getHomeViewController(this->navigationController);
    [homeVC updateWorkoutsList];
}

File: HealthApp/Tabs/Home/Main/Coordinator/HomeTabCoordinator.c
#include "HomeTabCoordinator.h"
#include <CoreGraphics/CGColor.h>
#include <dispatch/queue.h>
#include <math.h>
#include "AddWorkoutCoordinator.h"
#include "AppUserData.h"
#include "HomeVC.h"
#include "SetupWorkoutModal.h"
#include "ViewControllerHelpers.h"

#define setFloat(_o, _cmd, _arg) (((void(*)(id,SEL,float))objc_msgSend)((_o), (_cmd), (_arg)))

extern id kCAEmitterLayerLine;

enum {
    CustomWorkoutIndexTestMax,
    CustomWorkoutIndexEndurance,
    CustomWorkoutIndexStrength,
    CustomWorkoutIndexSE,
    CustomWorkoutIndexHIC
};

static void navigateToAddWorkout(HomeTabCoordinator *this, bool dismissVC, Workout *workout) {
    HomeTabCoordinator *coord = this;
    AddWorkoutCoordinator *child = malloc(sizeof(AddWorkoutCoordinator));
    child->navVC = this->navVC;
    child->parent = this;
    child->workout = workout;

    if (dismissVC) {
        dismissPresentedVC(getFirstVC(this->navVC), ^{
            coord->childCoordinator = child;
            addWorkoutCoordinator_start(child);
        });
    } else {
        coord->childCoordinator = child;
        addWorkoutCoordinator_start(child);
    }
}

static void showConfetti(id vc) {
    id view = getView(vc);
    CGRect frame;
    getRect(view, &frame, 0);
    id confettiView = createObjectWithFrame(objc_getClass("UIView"), frame);
    id grayColor = createColor(ColorGray);
    id bg = getObjectWithFloat(grayColor, sel_getUid("colorWithAlphaComponent:"), 0.8);
    setBackground(confettiView, bg);

    Class cellClass = objc_getClass("CAEmitterCell");
    SEL cgImg = sel_getUid("CGImage"), cgCol = sel_getUid("CGColor"), init = sel_getUid("init");
    SEL sColor = sel_getUid("setColor:"), sImg = sel_getUid("setContents:");
    SEL sRate = sel_getUid("setBirthRate:"), sLife = sel_getUid("setLifetime:");
    SEL sVel = sel_getUid("setVelocity:"), sELong = sel_getUid("setEmissionLongitude:");
    SEL sERange = sel_getUid("setEmissionRange:"), sSpin = sel_getUid("setSpin:");
    SEL sScale = sel_getUid("setScale:"), sScaleR = sel_getUid("setScaleRange:");

    int const velocities[] = {100, 90, 150, 200};
    id _colors[] = {
        createColor(ColorRed), createColor(ColorBlue),
        createColor(ColorGreen), getColorRef(1, 0.84, 0.04, 1)
    };
    CGColorRef shapeColors[4];
    id images[4];
    for (int i = 0; i < 4; ++i) {
        CFStringRef name = CFStringCreateWithFormat(NULL, NULL, CFSTR("confetti%d"), i);
        id img = createImage(name);
        CFRelease(name);
        images[i] = (id)((CGImageRef(*)(id,SEL))objc_msgSend)(img, cgImg);
        shapeColors[i] = ((CGColorRef(*)(id,SEL))objc_msgSend)(_colors[i], cgCol);
    }

    id cells[16];
    for (int i = 0; i < 16; ++i) {
        cells[i] = getObject(allocClass(cellClass), init);
        setFloat(cells[i], sRate, 4);
        setFloat(cells[i], sLife, 14);
        int velocity = velocities[arc4random_uniform(4)];
        setCGFloat(cells[i], sVel, velocity);
        setCGFloat(cells[i], sELong, M_PI);
        setCGFloat(cells[i], sERange, 0.5);
        setCGFloat(cells[i], sSpin, 3.5);
        ((void(*)(id,SEL,CGColorRef))objc_msgSend)(cells[i], sColor, shapeColors[i >> 2]);
        setObject(cells[i], sImg, images[i % 4]);
        setCGFloat(cells[i], sScaleR, 0.25);
        setCGFloat(cells[i], sScale, 0.1);
    }

    CFArrayRef array = CFArrayCreate(NULL, (const void **)cells, 16, &retainedArrCallbacks);
    id _layer = allocClass(objc_getClass("CAEmitterLayer"));
    id particleLayer = getObject(_layer, init);
    id viewLayer = getLayer(confettiView);
    setObject(viewLayer, sel_getUid("addSublayer:"), particleLayer);
    ((void(*)(id,SEL,CGPoint))objc_msgSend)(particleLayer, sel_getUid("setEmitterPosition:"),
                                            (CGPoint){frame.size.width / 2, 0});
    setObject(particleLayer, sel_getUid("setEmitterShape:"), kCAEmitterLayerLine);
    ((void(*)(id,SEL,CGSize))objc_msgSend)(particleLayer, sel_getUid("setEmitterSize:"),
                                           (CGSize){frame.size.width - 16, 1});
    setArray(particleLayer, sel_getUid("setEmitterCells:"), array);

    CFRelease(array);
    releaseObj(particleLayer);
    for (int i = 0; i < 16; ++i) {
        releaseObj(cells[i]);
    }
    addSubview(view, confettiView);

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, NSEC_PER_SEC * 5),
                   dispatch_get_main_queue(), ^(void) {
        removeView(confettiView);
        releaseObj(confettiView);
        id ctrl = createAlertController(localize(CFSTR("homeAlertTitle")),
                                        localize(CFSTR("homeAlertMessage")));
        addAlertAction(ctrl, localize(CFSTR("ok")), 0, NULL);
        presentVC(vc, ctrl);
    });
}

void homeCoordinator_start(HomeTabCoordinator *this) {
    fillStringArray(this->model.stateNames, CFSTR("homeState%d"), 2);
    fillStringArray(this->model.timeNames, CFSTR("timesOfDay%d"), 3);
    homeViewModel_updateTimeOfDay(&this->model);
    setupNavVC(this->navVC, homeVC_init(this));
}

void homeCoordinator_didFinishAddingWorkout(HomeTabCoordinator *this,
                                            int totalCompleted, bool popStack) {
    id homeVC = getFirstVC(this->navVC);
    homeVC_updateWorkoutsList(homeVC);

    if (popStack) {
        setBool(this->navVC, sel_getUid("popViewControllerAnimated:"), true);
        this->childCoordinator = NULL;
    }

    if (!totalCompleted) return;

    int nWorkouts = 0;
    for (int i = 0; i < 7; ++i) {
        if (this->model.workoutNames[i])
            ++nWorkouts;
    }

    if (nWorkouts == totalCompleted) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, NSEC_PER_SEC * 0.75),
                       dispatch_get_main_queue(), ^(void) {
            showConfetti(homeVC);
        });
    }
}

void homeCoordinator_addWorkoutFromPlan(HomeTabCoordinator *this, int index) {
    Workout *w = exerciseManager_getWeeklyWorkoutAtIndex(userData->currentPlan,
                                                         appUserData_getWeekInPlan(), index);
    if (w)
        navigateToAddWorkout(this, false, w);
}

void homeCoordinator_addWorkoutFromCustomButton(HomeTabCoordinator *this, int index) {
    unsigned char type = WorkoutStrength;
    switch (index) {
        case CustomWorkoutIndexSE:
            type = WorkoutSE;
            break;
        case CustomWorkoutIndexHIC:
            type = WorkoutHIC;
            break;
        case CustomWorkoutIndexTestMax: ;
            WorkoutParams params = {-1, WorkoutStrength, 2, 1, 1, 100};
            Workout *w = exerciseManager_getWorkoutFromLibrary(&params);
            if (w)
                navigateToAddWorkout(this, false, w);
            return;
        case CustomWorkoutIndexEndurance:
            type = WorkoutEndurance;
        default:
            break;
    }

    CFArrayRef names = exerciseManager_createWorkoutNames(type);
    if (!names) return;
    else if (!CFArrayGetCount(names)) {
        CFRelease(names);
        return;
    }

    presentModalVC(getFirstVC(this->navVC), setupWorkoutVC_init(this, type, names));
}

void homeCoordinator_finishedSettingUpCustomWorkout(HomeTabCoordinator *this, void *params) {
    Workout *w = exerciseManager_getWorkoutFromLibrary(params);
    if (w)
        navigateToAddWorkout(this, true, w);
}

void homeCoordinator_checkForChildCoordinator(HomeTabCoordinator *this) {
    AddWorkoutCoordinator *child = this->childCoordinator;
    if (child) {
        addWorkoutCoordinator_stopWorkoutFromBackButtonPress(child);
        this->childCoordinator = NULL;
    }
}
